File: 백준/Gold/16946. 벽 부수고 이동하기 4/벽 부수고 이동하기 4.py
import sys
input = lambda : sys.stdin.readline().rstrip()
n,m = map(int,input().split())
space = [input() for i in range(n)]
parent = list(range(n*m))
rank = [1]*n*m
def union(a,b):
    ra,rb = find(a),find(b)
    if ra==rb:
        return
    x,y = min(ra,rb), max(ra,rb)
    parent[y]=x
    find(a)
    find(b)

def find(x):
    if x!=parent[x]:
        # print(x,'rank update')
        rank[parent[x]]+=rank[x]
        rank[x]=0
        parent[x]=find(parent[x])
    return parent[x]
    
        

for i in range(n):
    for j in range(m):
        if space[i][j]=='1':
            continue
        index = m*i+j
        if j>0 and space[i][j-1]=='0':
            union(index-1,index)
            # print('left',i,j, rank)
            # print(parent)
        if i>0 and space[i-1][j]=='0':
            union(index-m,index)
            # print('up  ',i,j, rank)
        
result = [[0]*m for i in range(n)]
direction = [[0,1],[1,0],[-1,0],[0,-1]]
for i in range(n):
    for j in range(m):
        if space[i][j]=='0':
            continue
        result[i][j]+=1
        visited=set()
        for a,b in direction:
            dx,dy = i+a, j+b
            if 0<=dx<n and 0<=dy<m and  space[dx][dy]=='0':
                index = dx*m+dy
                pkey = find(index)
                if pkey in visited:
                    continue
                visited.add(pkey)
                result[i][j]+=rank[pkey]
                # print('index',[i,j],'+= rank',[dx,dy],rank[index])
        result[i][j]%=10
for i in result:
    print(*i, sep='')

File: 백준/Gold/16946. 벽 부수고 이동하기 4/벽 부수고 이동하기 4.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, m;
    static String[] space;
    static int[] parent, rank;
    static int[][] result;
    static int[][] direction = {{0,1},{1,0},{-1,0},{0,-1}};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        space = new String[n];
        for (int i = 0; i < n; i++) {
            space[i] = br.readLine();
        }

        parent = new int[n * m];
        rank = new int[n * m];
        for (int i = 0; i < n * m; i++) {
            parent[i] = i;
            rank[i] = 1;
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (space[i].charAt(j) == '1') continue;
                int index = i * m + j;
                if (j > 0 && space[i].charAt(j - 1) == '0') {
                    union(index - 1, index);
                }
                if (i > 0 && space[i-1].charAt(j) == '0') {
                    union(index - m, index);
                }
            }
        }

        result = new int[n][m];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (space[i].charAt(j) == '0') continue;
                result[i][j]++;
                HashSet<Integer> visited = new HashSet<>();
                for (int[] d : direction) {
                    int dx = i + d[0];
                    int dy = j + d[1];
                    if (0 <= dx && dx < n && 0 <= dy && dy < m && space[dx].charAt(dy) == '0') {
                        int index = dx * m + dy;
                        int pkey = find(index);
                        if (visited.contains(pkey)) continue;
                        visited.add(pkey);
                        result[i][j] += rank[pkey];
                    }
                }
                result[i][j] %= 10;
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int[] row : result) {
            for (int val : row) {
                sb.append(val);
            }
            sb.append('\n');
        }
        System.out.print(sb);
    }

    static int find(int x) {
        if (parent[x] != x) {
            int root = find(parent[x]);
            rank[root] += rank[x];
            rank[x] = 0;
            parent[x] = root;
        }
        return parent[x];
    }

    static void union(int a, int b) {
        int ra = find(a);
        int rb = find(b);
        if (ra == rb) return;
        int x = Math.min(ra, rb);
        int y = Math.max(ra, rb);
        parent[y] = x;
        find(a);
        find(b);
    }
}
